File: 프로그래머스/2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.js

// 정수 n을 k진법으로 변환하기
const getNdecimal = (n, k) => {
    let result = [];
        
    while (n >= k){
        result.push(n % k);
        n = Math.floor(n / k);
    }
    
    result.push(n);
    
    return result.reverse().join('');
}

// 일단 루트 n까지 나눠보자,,
const isPrime = (number) => {
    if (number === 0 || number === 1) return false;
    
    for (let i = 2; i * i <= number; i++){
        if (number % i === 0) {
            return false;
        }
    }
    
    return true;
}

function solution(n, k) {
    let answer = 0;
    
    // 정수 n을 k로 변환한다.
    let converted = getNdecimal(n, k);
    
    // 해당 변환한 수를 0으로 파싱한다.
    const numbers = converted.split('0').filter(num => num !== '');
    
    // 파싱한 수가 소수인지 판단한다.
    numbers.forEach(number => {
        if (isPrime(parseInt(number))) answer += 1;
    });
    
    return answer;
}

File: 프로그래머스/2/92335. k진수에서 소수 개수 구하기/k진수에서 소수 개수 구하기.java
import java.util.*;

class Solution {
    
    Set<Long> memo = new HashSet<>(); // 이미 등장한 소수 메모 
    public int solution(int n, int k) {
        
        int answer = 0;
        String[] splited = convertToBase(n, k).split("0");
        
        for (String number: splited) {
            if ("".equals(number)) {
                continue;
            }
            
            long num = Long.parseLong(number);
            
            if (memo.contains(num)) {
                answer++;
            } else {
                if (isPrimeNumber(num)) { // 소수인 경우
                    answer++;
                    memo.add(num);
                }
            }
        }
       
        return answer;
    }
    
    boolean isPrimeNumber(long num) {
        if (num <= 1) {
            return false;
        }
        
        for (long i = 2; i * i <= num; i++) {
            if (num % i == 0) {
                return false;
            }
        }
        return true;
    }
    
    // n을 k진법으로 바꾸는 함수
    String convertToBase(int n, int k) {
        StringBuilder sb = new StringBuilder();
        
        while(n > 0) {
            sb.append(n % k);
            n /= k;
        }
        
        return sb.reverse().toString();
    }
    
    void sout(Object o) {
        System.out.println(o);
    }
}